rwkv.cpp

```diff
@@ -84,72 +129,241 @@ inline enum rwkv_error_flags operator|=(enum rwkv_error_flags & a, enum rwkv_err
+}
+
+// Writes a single uint32 value to a file.
+bool fwrite_uint32(FILE * file, const uint32_t value) {
+    return fwrite((const void *) &value, sizeof(uint32_t), 1, file);
+}
+
+// Writes a single string value to a file.
+bool fwrite_string(FILE * file, const std::string & value) {
+    return fwrite((const void *) value.data(), value.length(), 1, file) == 1;
+}
+
+// Writes a single data buffer to a file.
+bool fwrite_data(FILE * file, const void * data, const size_t length) {
+    return fwrite(data, length, 1, file) == 1;
+}
+
+// --- File data structures ---
+
+#define TYPE_UNKNOWN TYPE_COUNT
+
+enum rwkv_type {
+    TYPE_F32,
+    TYPE_F16,
+    TYPE_Q4_0,
+    TYPE_Q4_1,
+    TYPE_Q4_1_O, // Unsupported
+    TYPE_Q4_2, // Unsupported
+    TYPE_Q4_3, // Unsupported
+    TYPE_Q5_0,
+    TYPE_Q5_1,
+    TYPE_Q8_0,
+    TYPE_COUNT
+};
+
+#define GGML_TYPE_UNKNOWN GGML_TYPE_COUNT
+
+static const enum ggml_type type_to_ggml[TYPE_COUNT + 1] = {
+    GGML_TYPE_F32,     /* F32    */
+    GGML_TYPE_F16,     /* F16    */
+    GGML_TYPE_Q4_0,    /* Q4_0   */
+    GGML_TYPE_Q4_1,    /* Q4_1   */
+    GGML_TYPE_UNKNOWN, /* Q4_1_O */
+    GGML_TYPE_UNKNOWN, /* Q4_2   */
+    GGML_TYPE_UNKNOWN, /* Q4_3   */
+    GGML_TYPE_Q5_0,    /* Q5_0   */
+    GGML_TYPE_Q5_1,    /* Q5_1   */
+    GGML_TYPE_Q8_0,    /* Q8_0   */
+    GGML_TYPE_COUNT    /* COUNT  */
+};
+
+static const enum rwkv_type type_from_ggml[GGML_TYPE_COUNT + 1] = {
+    TYPE_F32,    /* F32   */
+    TYPE_F16,    /* F16   */
+    TYPE_Q4_0,   /* Q4_0  */
+    TYPE_Q4_1,   /* Q4_1  */
+    TYPE_Q4_2,   /* Q4_2  */
+    TYPE_Q4_3,   /* Q4_3  */
+    TYPE_Q5_0,   /* Q5_0  */
+    TYPE_Q5_1,   /* Q5_1  */
+    TYPE_Q8_0,   /* Q8_0  */
+    TYPE_COUNT,  /* Q8_1  */
+    TYPE_COUNT,  /* I8    */
+    TYPE_COUNT,  /* I16   */
+    TYPE_COUNT,  /* I32   */
+    TYPE_COUNT,  /* COUNT */
+};
+
+static const char * type_to_string[TYPE_COUNT + 1] = {"float32", "float16", "Q4_0", "Q4_1", "Q4_1_O", "Q4_2", "Q4_3", "Q5_0", "Q5_1", "Q8_0", "unknown"};
+
+enum rwkv_type type_from_string(const char * str) {
+    for (int ord = 0; ord < TYPE_COUNT; ord++)
+        if (strcmp(str, type_to_string[ord]) == 0)
+            return (enum rwkv_type) ord;
+
+    return TYPE_UNKNOWN;
+}
+
+struct file_header {
+    uint32_t magic;
+    uint32_t version;
+    uint32_t n_vocab;
+    uint32_t n_embed;
+    uint32_t n_layer;
+    uint32_t data_type;
+};
+
+bool is_file_version_in_range(uint32_t version) {
+    return version >= RWKV_FILE_VERSION_MIN && version <= RWKV_FILE_VERSION_MAX;
+}
+
+bool fread_file_header(FILE * file, struct file_header & header, bool verify_data_type = true) {
+    RWKV_ASSERT_FALSE(RWKV_ERROR_FILE_READ, fread_data(file, sizeof(struct file_header), &header));
+    RWKV_ASSERT_FALSE(RWKV_ERROR_FILE_MAGIC, header.magic == RWKV_FILE_MAGIC);
+    RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_FILE_VERSION, is_file_version_in_range(header.version), "unsupported file version %" PRId32, header.version);
```

Comment:
```suggestion
    RWKV_ASSERT_FALSE_MSG(RWKV_ERROR_FILE_VERSION, is_file_version_in_range(header.version), "Unsupported file version %" PRId32, header.version);
```
Most messages start with upper-case letters, let's keep them consistent

Reply:
I actually tried to make them all consistently lowercase but ok